lib/core/src/Cardano/Pool/Metrics.hs

```diff
@@ -282,14 +282,24 @@ readPoolsPerformances
     -> SlotId
     -> m (Map PoolId Double)
 readPoolsPerformances DBLayer{..} (EpochLength el) tip = do
-    let range = [max 0 (currentEpoch - 14) .. currentEpoch]
-    atomically $ fmap avg $ forM range $ \ep -> calculatePerformance
+    atomically $ fmap avg $ forM historicalEpochs $ \ep -> calculatePerformance
         (slotsInEpoch ep)
         <$> (Map.fromList <$> readStakeDistribution ep)
         <*> (count <$> readPoolProduction ep)
   where
     currentEpoch = tip ^. #epochNumber
 
+    historicalEpochs :: [EpochNo]
+    historicalEpochs = map toEp
+        -- Have to be careful about underflow here!
+        [max 0 ((toInt currentEpoch) - 14) .. (toInt currentEpoch)]
+      where
+        toInt :: EpochNo -> Int
+        toInt (EpochNo e) = fromIntegral e
+
+        toEp :: Int -> EpochNo
+        toEp = EpochNo . fromIntegral
```

Comment:
Alternatively:

```hs
historicalEpochs :: [EpochNo]
historyEpochs 
    | currentEpoch < 14 -> [0..currentEpoch]
    | otherwise         -> [currentEpoch - 14 .. currentEpoch]
```

?

Reply:
Neat, although I think is more difficult to see that it is indeed correct.

Otherwise
```haskell
    historicalEpochs :: [EpochNo]
    historicalEpochs
        | currentEpoch < window = [0..currentEpoch]
        | otherwise             = [currentEpoch - window .. currentEpoch]
      where
        window = 14
```

```haskell
    historicalEpochs :: [EpochNo]
    historicalEpochs
        | currentEpoch > window = [currentEpoch - window .. currentEpoch]
        | otherwise             = [0..currentEpoch]
      where
        window = 14
```

but it still kind of relies on you knowing that `currenEpoch` and `window` are non-negative… but the same can probably be argued for the original `min` solution.

Reply:
Hmmm. I am not sure to get where you are going to with this? We know `currentEpoch` is non-negative because it's an `EpochNo`.